src/app/badge-generator.ts

```diff
@@ -0,0 +1,142 @@
+        const info = (repositoryField == null) ? null : this.inputInfo(repositoryField)
+        const source = this.source
+        if (info == null || option == "" || source == null) {
+            this.dom.preview()?.hide()
+            return;
+        }
+
+        this.client.repositoryInfo(info.owner, info.repo).then(repository => {
+            this.resetCopyButton()
+            this.dom.preview()?.show()
+            this.updatePreview(repository, option, info.owner, info.repo)
+        }).catch(reason => {
+            this.dom.preview()?.hide()
+            console.log(reason)
+        })
+    }
+
+    public onSelectOption(element: HTMLElement, event: Event) {
+        const source = this.source
+        if (source == null) return;
+
+        this.dom.preview()?.hide()
+        this.resetOptions()
+        this.updateOptions(source)
+    }
+
+    public onClickCopy(element: HTMLButtonElement, event: Event) {
+        this.dom.textArea()?.select()
+        this.dom.copySelection()
+        this.dom.textArea()?.setSelectionRange(0, 0)
+        this.setCopiedButton()
+    }
+
+    // helpers
+    private inputInfo(element: HTMLInputElement): (GitHubInput | null) {
+        const elements = element.value.replace(".git", "").split("/")
+        if (elements.length != 2) { return null; }
+        return new GitHubInput(elements[0], elements[1])
+    }
+
+    private resetCopyButton() {
+        this.dom.setCopyValue("Copy")
+    }
+
+    private setCopiedButton() {
+        this.dom.setCopyValue("Copied!")
+    }
+
+    private resetOptions() {
+        this.source = null
+
+        const selector = this.dom.sourceSelector()
+        selector?.removeChildren()
+
+        const option = document.createElement("option")
+        option.value = ""
+        option.innerText = "Choose..."
+        selector?.append(option)
+    }
+
+    private updateOptions(source: GitHubRepositorySource) {
+        this.source = source
+        if (source.tags.length < 0 && source.branches.length < 0) return;
+
+        const selector = this.dom.sourceSelector()
+        const options = this.dom.isTagSelected() ? source.tags : source.branches
+
+        options.forEach(option => {
+            const element = document.createElement("option")
+            element.value = option
+            element.innerText = option
+            selector?.append(element)
+        })
+    }
+
+    private updatePreview(info: GitHubRepository, option: string, owner: string, repo: string) {
+        const deeplink = this.deeplink(info, option, owner, repo)
+        
+        this.dom.setTagBranchName(option)
+        this.dom.setName(info.name)
+        this.dom.setAvatar(info.avatar)
+        this.dom.setDescription(info.description)
+        this.dom.setOwner(info.owner)
+        this.dom.setTextArea(`<a href="${deeplink}">${this.badge(info.name)}</a>`)
+    }
+
+    private deeplink(info: GitHubRepository, option: string, owner: string, repo: string): string {
```

Comment:
`option` here lets you pass whatever you want. If you model it as `Branch | Tag` it becomes more restrictive.